libopae++/samples/hello_fpga-1.cpp

```diff
@@ -0,0 +1,131 @@
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+#include "opae/cxx/properties.h"
+#include "opae/cxx/token.h"
+#include "opae/cxx/handle.h"
+#include "opae/cxx/dma_buffer.h"
+#include <uuid/uuid.h>
+#include <algorithm>
+#include <iostream>
+#include <chrono>
+
+using namespace opae::fpga::types;
+
+#ifndef LOG2_CL
+# define LOG2_CL 6
+#endif // LOG2_CL
+
+static uint64_t cacheline_aligned_addr(uint64_t num) {
+  return num >> LOG2_CL;
+}
+
+uint64_t operator "" _CL(unsigned long long int num) {
+  return num * 64;
+}
+
+size_t operator "" _KB(unsigned long long int num) {
+  return 1024 * num;
+}
+
+size_t operator "" _MB(unsigned long long int num) {
+  return 1024 * 1024 * num;
+}
+
+static const char* NLB0_AFUID = "D8424DC4-A4A3-C413-F89E-433683F9040B";
+
+static const size_t LPBK1_DSM_SIZE                 = 4_KB;
+static const size_t LPBK1_BUFFER_SIZE              = 1_KB;
+static const size_t LPBK1_BUFFER_ALLOCATION_SIZE   = 4_KB;
+static const uint64_t CSR_SRC_ADDR                 = 0x0120;
+static const uint32_t CSR_DST_ADDR                 = 0x0128;
+static const uint32_t CSR_CTL                      = 0x0138;
+static const uint32_t CSR_CFG                      = 0x0140;
+static const uint32_t CSR_NUM_LINES                = 0x0130;
+static const uint32_t DSM_STATUS_TEST_COMPLETE     = 0x40;
+static const uint64_t CSR_AFU_DSM_BASEL            = 0x0110;
+static const uint64_t CSR_AFU_DSM_BASEH            = 0x0114;
+
+int main(int argc, char* argv[]) {
+  // look for accelerator with NLB0_AFUID
+  properties filter;
+  filter.guid.parse(NLB0_AFUID);
+  filter.type = FPGA_ACCELERATOR;
+
+  auto tokens = token::enumerate({ filter });
+  
+  // assert we have found at least one
+  if (tokens.size() < 1) {
+  	std::cerr << "accelerator not found\n";
+  	return -1;
+  }
+  token::ptr_t tok = tokens[0];
+
+  // open accelerator and map MMIO
+  auto accel = handle::open(tok, FPGA_OPEN_SHARED);
+
+  // allocate buffers
+  auto dsm = dma_buffer::allocate(accel, LPBK1_DSM_SIZE);
+  auto inp = dma_buffer::allocate(accel, LPBK1_BUFFER_ALLOCATION_SIZE);
+  auto out = dma_buffer::allocate(accel, LPBK1_BUFFER_ALLOCATION_SIZE);
+
+  // initialzie buffers
```

Comment:
initialzie -> initialize

I don't like the dma_buffer::fill method. The data size is ambiguous. It takes an uint32_t and passes it to memset, so it's a byte value? Why should we have so many "helper" functions in the base dma_buffer class (like fill() and equal())?

Reply:
Thanks for your feedback. Helper functions was something that had been up for debate too. We started putting helper functions in another file to track them. We were thinking this other file would probably be part of the second layer so it shouldn't be considered as part of this first layer. The file can be found here for now (in feature/c++ branch): `common/include/opae/cxx/buffers.h`
Perhaps we'll move `fill` and `equal` to there too.

Reply:
I wonder if we should leave them in dma_buffer.  My thinking is that the base layer should provide what is necessary to perform a minimal AFU interaction (Lpbk1). In that case, if we removed these, then the user would need to acquire the raw buffer pointer and do the initialization / comparison on their own.

Reply:
I think a helper function in dma_buffer to zero the buffer is perfectly reasonable. It's something likely to be common across applications. Setting a buffer to a repeated value of a particular byte is unlikely to be universal and the semantics of the call are not at all clear, since it isn't obvious that fill is byte sized. I don't think you should be catering to the needs of lpbk1 in the class.